[vcpkg] Allow CI to pass in all relevant directories and remove use of symbolic links

BillyONeal opened the pull request:
Adds `--x-foo-root=` parameters for the directories to which we have applied symlinks in the past, such as installed, buildtrees, packages, etc.

This is intended to unblock https://github.com/microsoft/vcpkg/pull/11365 which broke due the symlink workaround we previously used to get the work tree onto the VM temp disk not working with the current Scale Set Agents installation script.

Drive-by fixes:
* Fixed help text having differing output and alignment caused by duplicating these declarations in the tool
![image](https://user-images.githubusercontent.com/1544943/82517869-ef7d5700-9ad2-11ea-87f2-7990ad4a88ab.png)
* Always use HelpTableFormatter to emit help tables.
* Fixed parameter validation, filesystem validation, and path conversion being spewed out in between VcpkgCmdArguments, VcpkgPaths, and main(). Now VcpkgCmdArguments owns command line argument parsing and reading defaults from the environment, VcpkgPaths owns only path resolution bits. Theoretically this reduces the fidelity of the error messages we emit here, but I don't think in practice customized error messages for each type of missing input or output directory add much user value. We emit messages containing the bad path whenever something explodes anyway.

BillyONeal committed:
Plumb passing around filesystem the right way.

BillyONeal committed:
Add parameter parsing for all the extra directories we want to allow being set in CI.

BillyONeal committed:
Rename variables of VcpkgCmdArguments to have correct underscores for consistency with the rest of the vcpkg codebase.

BillyONeal committed:
Always use display_usage for options printing. (This is duplicating some bits right now, but less than before)

BillyONeal committed:
Promote the help table formatter to vcpkgcommandarguments, since that formatting is really about console access which is an arguments thing, not a help command specific thing.

Use the help table formatter for all help text.

BillyONeal committed:
Restore blank lines from the old form.

BillyONeal committed:
Reduce target width to 100 characters (from 120)

BillyONeal committed:
Move all filesystem bits into VcpkgPaths::VcpkgPaths and wire up all the new options like install root etc.

BillyONeal committed:
Plumb some of these directories though vcpkg_configure_cmake.

BillyONeal committed:
Hook up Linux CI without symlinks.

BillyONeal committed:
Hook up Windows CI without symlinks.

BillyONeal committed:
Hook up MacOS CI without symlinks.

BillyONeal committed:
Ignore when the directories already exist for Linux.

Neumann-A commented on scripts/buildsystems/vcpkg.cmake:
```diff
@@ -151,7 +151,9 @@ if(NOT DEFINED _VCPKG_ROOT_DIR)
     set(_VCPKG_ROOT_DIR ${_VCPKG_ROOT_DIR_CANDIDATE} CACHE INTERNAL "Vcpkg root directory")
 endif()
-set(_VCPKG_INSTALLED_DIR ${_VCPKG_ROOT_DIR}/installed)
+if (NOT DEFINED _VCPKG_INSTALLED_DIR)
+    set(_VCPKG_INSTALLED_DIR ${_VCPKG_ROOT_DIR}/installed)
+endif()
```
Hmm is there a reason to not directly pass `installed/triplet` instead?

Neumann-A replied:
Probably should replace all usage of `${_VCPKG_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}` to `${VCPKG_INSTALLED_TRIPLET_DIR}` in the toolchain `vcpkg.cmake` and the wrappers `vcpkg_cmake_wrappers.cmake`

BillyONeal replied:
This is the file users who use vcpkg include, so we don't know what the triplet is yet. We could pass that in rather than detecting it too; I was trying to make minimal changes to the cmake infrastructure as I understand cmake less than c++ :)

BillyONeal replied:
Can do that

BillyONeal replied:
On second thought, this change is already huge enough, can that be done in a followup PR?

Neumann-A replied:
Do as you like.

Voskrese commented:
merge?

BillyONeal commented:
Depends on https://github.com/microsoft/vcpkg/pull/11483 and https://github.com/microsoft/vcpkg/pull/11545

BillyONeal committed:
Fix failure log uploading.

BillyONeal committed:
Fix msbuild projects that USE_VCPKG_INTEGRATION

Neumann-A commented on scripts/cmake/vcpkg_build_msbuild.cmake:
```diff
@@ -109,7 +109,7 @@ function(vcpkg_build_msbuild)
             APPEND _csc_OPTIONS
             /p:ForceImportBeforeCppTargets=${SCRIPTS}/buildsystems/msbuild/vcpkg.targets
-            "/p:VcpkgTriplet=${TARGET_TRIPLET}"
+            "/p:VcpkgRoot=${CURRENT_INSTALLED_DIR}/${VCPKG_TARGET_TRIPLET}"
```
That is wrong.
You probably want
```
"/p:VcpkgTriplet=${TARGET_TRIPLET}"
"/p:VcpkgRoot=${VCPKG_ROOT_DIR}"
```

Neumann-A replied:
`VcpkgRoot` should be renamed to `VcpkgInstalledTriplet` instead....

BillyONeal replied:
I agree :)

strega-nil replied:
Is `${CURRENT_INSTALLED_DIR}` `installed/triplet` or just `installed`?

BillyONeal committed:
Fix ports improperly naming ${VCPKG_ROOT_DIR}/installed

strega-nil commented on scripts/azure-pipelines/linux/azure-pipelines.yml:
```diff
@@ -10,42 +10,40 @@ jobs:
-    displayName: 'Report on Disk Space After Environment'
-    condition: always()
+      sudo mkdir /mnt/vcpkg-ci -m=777
+      sudo mkdir /mnt/vcpkg-ci/downloads -m=777
```
(Given this without an intervening `mount`, this seems like it's just on the local hard drive? which is fine, but why's it in `/mnt`)

strega-nil commented on ports/spirv-tools/portfile.cmake:
```diff
@@ -20,7 +20,7 @@ vcpkg_configure_cmake(
     PREFER_NINJA
     OPTIONS
-        -DSPIRV-Headers_SOURCE_DIR=${VCPKG_ROOT_DIR}/installed/${TARGET_TRIPLET}
+        -DSPIRV-Headers_SOURCE_DIR=${CURRENT_INSTALLED_DIR}
```
Because here, you've got `${CURRENT_INSTALLED_DIR}` being used to mean `installed/triplet`, afaict.

BillyONeal replied:
`set(CURRENT_INSTALLED_DIR ${_VCPKG_INSTALLED_DIR}/${TARGET_TRIPLET} CACHE PATH "Location to install final packages")`

No I don't know why the spirv build succeeds ¯\_(ツ)_/¯

strega-nil replied:
So then this use is correct, and it should succeed.

BillyONeal committed:
Fix powershell syntax error.

strega-nil reviewed (changes requested):
There's some weird stuff, but it's mostly minor. Thanks @BillyONeal :)

BillyONeal commented on scripts/azure-pipelines/linux/azure-pipelines.yml:
```diff
@@ -10,30 +10,27 @@ jobs:
-      df -h
-    displayName: 'Report on Disk Space After Environment'
-    condition: always()
+      sudo mkdir /mnt/vcpkg-ci -m=777
```
Azure mounts the temporary disk to `/mnt`; see the last line of df -h in one of our build logs:

```
/bin/bash --noprofile --norc /agent/_work/_temp/6c461458-6559-418b-a4db-89d8303fb4fc.sh
Filesystem                                      Size  Used Avail Use% Mounted on
udev                                             32G     0   32G   0% /dev
tmpfs                                           6.3G  700K  6.3G   1% /run
/dev/sdb1                                        29G  5.6G   24G  20% /
tmpfs                                            32G  8.0K   32G   1% /dev/shm
tmpfs                                           5.0M     0  5.0M   0% /run/lock
tmpfs                                            32G     0   32G   0% /sys/fs/cgroup
/dev/sdb15                                      105M  3.6M  101M   4% /boot/efi
//prlin20200507.file.core.windows.net/archives  2.0T  1.1T  953G  54% /archives
/dev/sda1                                       393G   26G  347G   7% /mnt
```

strega-nil replied:
... I hate it.

BillyONeal replied:
I don't think there's anything we can change it to that would make it obvious but I can add a comment.

strega-nil replied:
that sounds good to me :)

BillyONeal commented on toolsrc/src/vcpkg/build.cpp:
```diff
@@ -449,22 +453,26 @@ namespace vcpkg::Build
                 hash += "-";
-                hash += Hash::get_file_hash(VCPKG_LINE_INFO, fs, paths.scripts / "toolchains" / "freebsd.cmake", algo);
+                hash += Hash::get_file_hash(
+                    VCPKG_LINE_INFO, fs, paths.scripts / fs::u8path("toolchains") / fs::u8path("freebsd.cmake"), algo);
```
The multiple operator/s ensure that the preferred separator is used there... I was being paranoid :)

strega-nil replied:
Given how quickly Windows converts `/` to `\`, I'm not worried about it. If you are, then you can change it.

BillyONeal replied:
Changed.

BillyONeal commented on scripts/azure-pipelines/windows/azure-pipelines.yml:
```diff
@@ -15,14 +15,16 @@ jobs:
     inputs:
       filePath: 'scripts/azure-pipelines/windows/initialize-environment.ps1'
-  - task: BatchScript@1
+  - task: CmdLine@2
```
That has different effects. As written this causes ADO to run that script directly, with the short form you show it creates a script file with the content `.\boostrap-vcpkg.bat`, and then invokes that temporary.

I guess it's not a huge deal either way.

strega-nil replied:
oh, then don't worry about it.

BillyONeal committed:
Go back to invoking directly as the encoding problem hasn't recurred recently.

BillyONeal committed:
Don't bother saying System.DefaultWorkingDirectory because that's CWD.

BillyONeal committed:
Add temporary disk comments.

BillyONeal committed:
Avoid double u8path as requested by Nicole in code review.

BillyONeal committed:
Fix places that were double-adding the triplet.

BillyONeal commented:
> There's some weird stuff, but it's mostly minor. Thanks @BillyONeal :)

idk I think you found a serious bug!

BillyONeal committed:
PowerShell still hates me.

BillyONeal commented:
/azp run

strega-nil commented on scripts/cmake/vcpkg_build_msbuild.cmake:
```diff
@@ -110,6 +110,7 @@ function(vcpkg_build_msbuild)
             APPEND _csc_OPTIONS
             /p:ForceImportBeforeCppTargets=${SCRIPTS}/buildsystems/msbuild/vcpkg.targets
             "/p:VcpkgTriplet=${TARGET_TRIPLET}"
+            "/p:VcpkgRoot=${CURRENT_INSTALLED_DIR}"
```
is `VcpkgRoot` the equivalent of `CURRENT_INSTALLED_DIR`? I would expect this variable to be something like where the `vcpkg` executable lives.

BillyONeal replied:
Yes, it is like `CURRENT_INSTALLED_DIR` -- the name is bad because I didn't write the script consuming the variable (it's `$/scripts/buildsystems/msbuild/vcpkg.targets`). I don't want to edit that here because I don't know exactly how it interacts with `vcpkg integrate install`. Would filing an issue about it and addressing in a subsequent PR be OK? (I don't want to grow this PR even bigger :) )

BillyONeal replied:
Hmmm I guess I could make a dependent change which changes that name, stand by...

BillyONeal replied:
https://github.com/microsoft/vcpkg/pull/11653

strega-nil replied:
so should we merge 11653 first?

BillyONeal replied:
Yes, done.

strega-nil commented on scripts/cmake/vcpkg_install_msbuild.cmake:
```diff
@@ -137,6 +137,7 @@ function(vcpkg_install_msbuild)
         /p:UseIntelMKL=No
         /p:WindowsTargetPlatformVersion=${_csc_TARGET_PLATFORM_VERSION}
         /p:VcpkgTriplet=${TARGET_TRIPLET}
+        "/p:VcpkgRoot=${CURRENT_INSTALLED_DIR}"
```
Same questions as with `build_msbuild`.

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into directories

# Conflicts:
#	toolsrc/include/vcpkg/vcpkgcmdarguments.h
#	toolsrc/src/vcpkg.cpp
#	toolsrc/src/vcpkgcmdarguments.cpp

BillyONeal committed:
Fix parse_switch calls.

BillyONeal committed:
Fix MSBuild variable name.

strega-nil commented on toolsrc/src/vcpkg/vcpkgcmdarguments.cpp:
```diff
@@ -460,13 +516,155 @@ namespace vcpkg
+    {
+        if (!disable_metrics)
+        {
+            const auto VCPKG_DISABLE_METRICS_ENV = System::get_environment_variable("VCPKG_DISABLE_METRICS");
```
Why are these all SCREAMING_SNAKES?

BillyONeal replied:
I wanted to indicate that the variable is in exactly the same as the environment variable of the same name (that's also the pattern below). I had to add an ENV suffix here though because `VCPKG_DISABLE_METRICS` is also already a macro :/

BillyONeal replied:
I am not wed to this convention, SUPER open to other suggestions :)

strega-nil replied:
I'd prefer `vcpkg_disable_metrics_env` style, personally.

BillyONeal replied:
Done.

BillyONeal committed:
Be less SHOUTY_SNAKES

JackBoosY reviewed (changes requested):
Please update `qt5-base`, `shiva-sfml`, `spirv-tools` version info. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#update-the-version-field-in-the-control-file-of-any-modified-ports).

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into directories

BillyONeal committed:
Bump version number on edited packages.

BillyONeal commented:
> Please update `qt5-base`, `shiva-sfml`, `spirv-tools` version info. See [documentation](https://github.com/microsoft/vcpkg/blob/master/docs/maintainers/maintainer-guide.md#update-the-version-field-in-the-control-file-of-any-modified-ports).

Done, thanks!

BillyONeal committed:
Defend against no trailing slash of VcpkgRoot and VcpkgCurrentInstalledDir.

I could have ensured we always passed the trailing slash, but defending against it here makes it easier for users if they want to use /p themselves.

Maximus5 commented on scripts/buildsystems/msbuild/vcpkg.targets:
```diff
@@ -54,7 +54,13 @@
+    allow us to override a console provided property here. -->
+    <VcpkgRootSanitized>$(VcpkgRoot)</VcpkgRootSanitized>
+    <VcpkgRootSanitized Condition="!$(VcpkgRootSanitized.EndsWith('\'))">$(VcpkgRootSanitized)\</VcpkgRootSanitized>
+    <VcpkgCurrentInstalledDirSanitized>$(VcpkgCurrentInstalledDir)</VcpkgCurrentInstalledDirSanitized>
```
Looks like there is a problem here.
The `VcpkgCurrentInstalledDir` initialization was removed few lines above and it's empty now.
